src/appleseed/renderer/kernel/shading/shadingpoint.cpp

```diff
@@ -442,7 +442,10 @@ void ShadingPoint::refine_and_offset() const
 
             assert(is_curve_primitive());
 
-            m_refine_space_geo_normal = normalize(-refine_space_ray.m_dir);
+            const Vector3d x = normalize(get_dpdu(0));
+            const Vector3d& sn_curve = normalize(-refine_space_ray.m_dir);
+            const Vector3d z = normalize(cross(sn_curve, x));
```

Comment:
I don't think you need `x`, `z` or `sn_curve` to be unit-length.

Reply:
Are you sure that `sn_curve ^ x` is correct? Ideally (x, sn_curve, z) would form a right-handed coordinate system, but here it doesn't. That said, a comment of yours further down seem to indicate that this may be correct and that curves use a different convention, for some reason.

Reply:
Yes, usually the coordinate system will be the of the form you mentioned, but the hair shader requires a shading coordinate system with the x and z axes along the fibre and perpendicular to the fibre(and viewing direction), respectively i.e the dpdu and dpdv components. The shading normal is then calculated from the cross product of both these components.